### ISSUE TITLE
Cadvisor and rkt api service race against each other

### ISSUE THREAD

**Issue by sjpotter** [google/cadvisor#1255]
In order to get cadvisor to work well with rkt we need to make use of the rkt api service.

We need the api service for a few reasons

1) the current rkt implementation depends on the rkt uuid being encoded in the cgroup name.  This isn't guaranteed.  It only happens if run as "rkt run" from the command line or if the user names their systemd unit file with the uuid.  encoding the UUID into the unit file is difficult as it excludes some use cases.

ex: imagine if one wants to run etcd or any other service in a rkt container on every boot - ala examples given at https://coreos.com/rkt/docs/latest/using-rkt-with-systemd.html  One can't encode the UUID into the file as it will change on every boot

2) without the rkt api service, we can't get the annotations that are set for this container, this is especially important for k8s as it depends on the annotations being returned as Cadvisor's Labels map to be able to internally map between cadvisor metrics for containers and what the kubelet thinks the containers are.

So to solve these things we need to be able to map a cgroup that cadvisor sees to the appropriate Pod struct that the api-service can return.

However, there is a race condition in terms of when the cgroup appears (and cadvisor sees it and tried to create a handler for it) and when the rkt api service can return data for it

There are a few ways to run a rkt pod

1) call rkt run from the command line
2) call rkt prepare ; rkt run-prepare from the command line
3) call rkt run from a systemd unit
4) call rkt prepare (from anywhere I think) and rkt run-prepare from a systemd unit

I've tested 1/3/4 and in case 1 the rkt pod cgroups are under machine.slice, while in case 3/4 they are under system.slice (by default, systemd allows you to specify what cgroup they should appear under and one can create their own)

In my testing I never saw a case where if I did rkt run from the command line that when the rkt api service was called frm cadvisor that it didn't know about the container, I hypothosise that this is probably because by the time systemd-nspawn is called, rkt views the pod as in the run state.

However, if run from a systemd unit, we do see races, as systemd create a new cgroup for the unit before rkt runs and the pod can be only in the prepared state (and not be findable via cgroups) or not in any state if its a rkt run.  

A solution I came up was to tag a "rkt" systemd units by telling the unit to put it into its own slice (which the using rkt with systemd instructions above seem to recommend itself) we can then know that we should busy wait until the rkt api service knows about this cgroup (or perhaps even tell the rkt api service to busy wait)

A solution others proposes is to tag it with some other metadata that can be queried via dbus to tell cadvisor that this is a rkt pod.  I'm not convinced there's much of a difference between these 2 approaches as in some cases we will still have to busy wait, we can perhaps minimize the need to busy wait in the cases of something prepared and we provide its uuid as metadata, but it wont solve all the cases.

@philips @yifan-gu @euank

**Comment by yifan-gu** [comment]
Try to think out loud here. So why should we watch cgroup fs to detect new containers in the first place?
Is there way to just talk to rkt api service directly to list containers and get new containers?

**Comment by yifan-gu** [comment]
As @jellonek pointed out here https://github.com/kubernetes/kubernetes/issues/23889#issuecomment-206382577 Relying on cgroups will not work for kvm stage1 anyway. And containers don't just mean cgroup + namespace. 
cAdvisor today does seem to have some abstraction (i.e. the [ContainerHandler](https://github.com/google/cadvisor/blob/v0.23.0/container/container.go#L48)) I am not sure if that's enough though. If we implement that interface without using cgroup, will the rkt + cAdvisor work?

**Comment by yifan-gu** [comment]
But back to @sjpotter 's specific problem here, the cause is when running rkt through systemd.service, we gave the nspawn `--keep-unit` flag, which instead of creating it's own cgroup, it uses the cgroup created by systemd.
> --keep-unit
           Instead of creating a transient scope unit to run the container in, simply register the service or scope unit systemd-nspawn has been
           invoked in with systemd-machined(8). This has no effect if --register=no is used. This switch should be used if systemd-nspawn is
           invoked from within a service unit, and the service unit's sole purpose is to run a single systemd-nspawn container. This option is not
           available if run from a user session.

Wonder if removing that flag is a feasible approach? cc @iaguis @alban @jonboulle

**Comment by timstclair** [comment]
> Relying on cgroups will not work for kvm stage1 anyway. And containers don't just mean cgroup + namespace. 
cAdvisor today does seem to have some abstraction (i.e. the ContainerHandler) I am not sure if that's enough though. If we implement that interface without directly talking with the cgroup fs, will the rkt + cAdvisor work?

cAdvisor still makes some assumptions around containers using cgroups, though we could refactor more of that logic into the ContainerHandler interface. Another problem is that if rkt is still generating cgroups, the [raw handler](https://github.com/google/cadvisor/blob/4bebcc0e1f24be0d810c174420e1a0ef85062fc4/container/raw/factory.go#L62) will pick it up (it's a catch-all for cgroups-based containers).

**Comment by sjpotter** [comment]
tim the last point doesn't bother me as we can easily get the rkt factory to say I could handle this container but I don't accept it and then it wont be handled by any handler.

**Comment by sjpotter** [comment]
I'd also guess that hyper and all VM based systems will have the same problem as rkt w/ kvm stage1

**Comment by iaguis** [comment]
> Wonder if removing that flag is a feasible approach? cc @iaguis @alban @jonboulle

If we remove that flag, the limits applied to the unit file won't apply to the processes in the pod, which is kind of a fundamental goal for rkt.

**Comment by sjpotter** [comment]
So in thinking about the code.

Currently we only have one "watcher", which is basically the inotify watcher provided by the raw cgroup handler as well as the ability to recursively list all cgroups to detect any that inotify might have missed

I don't see a reason why we can't register watchers (all they are are go routines) that would be appropriate for each runtime much like we register multiple container handlers themselves.

each handler would know how to handle its own watch events as well as return a !accept for cgroups that correspond to its containers (which it will handle via its own mechanism)

however, you still have the race problem of being able to determine if a cgroup belongs to your container or not (i.e. the rkt race described above).  

the other thing to think about is that perhaps it doesn't matter if there's a raw handler and a specific handler for the same container, as I'm not sure anything the raw handler does uses much cpu (i.e. the rkt/docker handlers are busier in terms of memory/cpu usage due to how they track disk usage with ```du``` but I'm not sure there's really much overhead from the raw handler itself.

so in this model, each watcher will not just pass a name of the container it sees, but also the "type" (which could be raw), handlers will know which types they support and ignore the ones they don't.  so multiple handlers might support raw cgroups names, but we could also have a rkt names, hyper names....

**Comment by sjpotter** [comment]
> > Wonder if removing that flag is a feasible approach? cc @iaguis @alban @jonboulle

> If we remove that flag, the limits applied to the unit file won't apply to the processes in the pod, which is kind of a fundamental goal for rkt.

so we can't set limits on the command line if we run "rkt run" its only settable if run as a systemd unit?

And there's no way to determine what the limits of the cgroup you are running in are?

**Comment by iaguis** [comment]
> so we can't set limits on the command line if we run "rkt run" its only settable if run as a systemd unit?

Not now, but it can be implemented. Note that if you pass a bigger value on the command line than the one in the "outer" service file, the value on the service file will be respected.

> And there's no way to determine what the limits of the cgroup you are running in are?

You can read the values of the controllers of your own cgroup hierarchy (check `/proc/self/cgroup`).

**Comment by sjpotter** [comment]
so wouldn't that be a reasonable way to go forward?

1) enable users to set limits on the rkt command line
2) enable one to inherit limits from the running context
3) if inheriting limits, don't allow them to be overridden to be greater.

**Comment by iaguis** [comment]
You mean non inheriting by default (that is, don't pass `--keep-unit`)?

> edit: also what do --cpu and --memory do now then?

They are per-app limits, not pod-level limits.

**Comment by sjpotter** [comment]
I mean manually inherit.  I.e. instead of running in the same unit, which i think is what waxing it does now copy the limits to a new cgroup and run in it.

**Comment by philips** [comment]
The wait is only annoying because we can't tell if a cgroup _should_ wait on rkt metadata. In the case of rkt it needs to wait; but in other cases it doesn't need to wait.

To fix this we should just put explicitly in a unit file that the type is rkt then we can have cAdvisor parse it.

For example: 

```
[Service]
ExecStart=/bin/sleep 3000

[X-cAdvisor]
Type=rkt
```

Working code to talk to systemd and grab this metadata:

https://gist.github.com/philips/07aae9eea7ffddd75cf0c1d6ecf839b1

**Comment by sjpotter** [comment]
@philips if my understanding of the issue is correct that the race is due to --keep-unit  and and one would only call rkt with --keep-unit within the context of a systemd unit that this might work.

I can experiment with it

**Comment by sjpotter** [comment]
So my current approach is to do major surgery on cadvisor and refactor out the watcher to make it easy to plug in a parallel rkt watcher that doesn't really on inotify on the cgroup paths

first part (refactorin cadvisor) is here https://github.com/google/cadvisor/pull/1263

second part that implements it for rkt is not PRd yet but is here https://github.com/sjpotter/cadvisor/tree/new-handler-detection-with-rkt/watcher/rkt

**Comment by sjpotter** [comment]
pretty sure I solved this issue with my refactor of the cadvisor/rkt integration last spring.
